node/actors/bft/src/testonly/run.rs

```diff
@@ -71,77 +110,407 @@ impl Test {
                 }
-                scope::run!(ctx, |ctx, s| async {
-                    for recv in recvs {
-                        s.spawn(async {
-                            use zksync_consensus_network::io;
-                            let mut recv = recv;
-                            while let Ok(io::InputMessage::Consensus(message)) =
-                                recv.recv(ctx).await
-                            {
-                                let msg = || {
-                                    io::OutputMessage::Consensus(io::ConsensusReq {
-                                        msg: message.message.clone(),
-                                        ack: oneshot::channel().0,
-                                    })
-                                };
-                                match message.recipient {
-                                    io::Target::Validator(v) => sends.get(&v).unwrap().send(msg()),
-                                    io::Target::Broadcast => {
-                                        sends.values().for_each(|s| s.send(msg()))
-                                    }
+            },
+            io::Target::Validator(v) => {
+                let target_ports = &validator_ports[&v];
+
+                match partitions_opt {
+                    None => {
+                        for target_port in target_ports {
+                            eprintln!(
+                                "unicasting view={view_number} from={port} target={target_port}"
+                            );
+                            send_or_stash(true, *target_port, msg());
+                        }
+                    }
+                    Some(ps) => {
+                        for p in ps {
+                            let can_send = p.contains(&port);
+                            for target_port in target_ports {
+                                if p.contains(target_port) {
+                                    eprintln!("unicasting view={view_number} from={port} target={target_port } can_send={can_send} t={}", start.elapsed().as_secs());
+                                    send_or_stash(can_send, *target_port, msg());
                                 }
                             }
-                            Ok(())
-                        });
+                        }
                     }
-                    anyhow::Ok(())
-                })
-                .await?;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Simulate the effects of gossip: if the source node can gossip to the target,
+/// and the message being sent contains a CommitQC, and the sender has the
+/// referenced finalized block in its store, then assume that the target
+/// could fetch this block if they wanted, and insert it directly into their store.
+///
+/// This happens concurrently with the actual message passing, so it's just an
+/// approximation. It also happens out of order. This method only contains the
```

Comment:
would it be hard to implement gossip by looking at the current partitions and syncing any missing blocks directly, rather than do syncing on connections induced by other messages?

Reply:
At the moment there is no concept of "current" in the simulation, the partitions are only consulted in reaction to a message being passed from one actor to another, and for that the view is determined from the message. This is why I left comments to remind that if ChonkyBFT stayed in the same view until a quorum is reached, and we created partitions which have less members than required for a quorum, then we would need another mechanism to lift the partitions in the view "after a while". I left that for a future exercise because I didn't want to introduce timeouts in this test, and the current implementation goes on to the next view on its own anyway.

If I understand your question correctly you're also asking if it would be better to have a sync loop periodically looking at pulling missing blocks, without being triggered by a different message being sent. 

I don't have a strong opinion on that. IIUC the way syncing works is it first waits to be notified via gossip that the remote peer has a range of blocks we are interested in, and then pulls the block via RPC, so a probing loop would still be just an approximation, as it would be missing the initial message from the "have" to the "have not". 

It would be great if the actual gossip messages could be part of the simulation itself, but short of that I don't see one way being much better than the other, as long as we don't introduce stronger assumptions than what is realistic, e.g. by always inserting into the target store before even sending the message, so the target never has to wait for missing blocks to appear.